sunscreen_math/src/metal_impl/mod.rs

```diff
@@ -0,0 +1,128 @@
+use std::mem::size_of;
+use std::ops::Deref;
+
+use lazy_static::lazy_static;
+use metal::{
+    Buffer, CommandQueue, Device, Library, MTLCommandBufferStatus, MTLResourceOptions, MTLSize,
+};
+
+mod ristrettovec;
+pub use ristrettovec::*;
+mod scalarvec;
+pub use scalarvec::*;
+
+// In build.rs, we compile 2 variants of the curve25519-dalek.metallib library:
+// test and release. In test, we #define the TEST macro, which exposes test kernels.
+// The release library does not feature these kernels.
+#[cfg(not(test))]
+const SHADERLIB: &[u8] = include_bytes!(concat!(
+    env!("OUT_DIR"),
+    "/curve25519-dalek.release.metallib"
+));
+
+#[cfg(test)]
+const SHADERLIB: &[u8] =
+    include_bytes!(concat!(env!("OUT_DIR"), "/curve25519-dalek.test.metallib"));
+
+pub struct Runtime {
+    device: Device,
+    lib: Library,
+    command_queue: CommandQueue,
+}
+
+/// TODO Remove
+unsafe impl Sync for Runtime {}
+
+lazy_static! {
+    static ref RUNTIME: Runtime = {
+        let device = Device::system_default().unwrap();
+        let lib = device.new_library_with_data(SHADERLIB).unwrap();
+        let command_queue = device.new_command_queue();
+
+        Runtime {
+            device: Device::system_default().unwrap(),
+            lib,
+            command_queue,
+        }
+    };
+}
+
+impl Runtime {
+    pub fn get() -> &'static Runtime {
+        &RUNTIME
+    }
+
+    pub fn alloc(&self, len: usize) -> Buffer {
+        self.device
+            .new_buffer(len as u64, MTLResourceOptions::StorageModeShared)
+    }
+
+    pub fn run(&self, kernel_name: &'static str, data: &[&Buffer], grid: [(u64, u64); 3]) {
+        println!("{}", self.device.recommended_max_working_set_size());
+
+        let command_buffer = self.command_queue.new_command_buffer();
+        let encoder = command_buffer.new_compute_command_encoder();
+
+        let gpu_fn = self.lib.get_function(kernel_name, None).unwrap();
+        let gpu_fn = self
+            .device
+            .new_compute_pipeline_state_with_function(&gpu_fn)
+            .unwrap();
+
+        dbg!(gpu_fn.max_total_threads_per_threadgroup());
+
+        encoder.set_compute_pipeline_state(&gpu_fn);
+
+        for (i, buf) in data.iter().enumerate() {
+            encoder.set_buffer(i as u64, Some(buf), 0);
+        }
+
+        let global = MTLSize::new(grid[0].0, grid[1].0, grid[2].0);
+        let local = MTLSize::new(grid[0].1, grid[1].1, grid[2].1);
```

Comment:
Do we reuse 3x3 grids often? It might be worth making a type alias (or a newtype) for the grid. Plus it is easier for the caller to know what the `[(u64, u64); 3]` represents then.

Reply:
👍